articles/node_lifecycle.md

```diff
@@ -0,0 +1,176 @@
+
+Any service requests to a node in the inactive state will not be answered (to the caller, they will fail immediately).
+
+Entering the `inactive` state from the `created` state requires running the node's `onInitialised` method. The node uses this to set up any resources it must hold throughout its life (irrespective of if it is active or inactive). As examples, such resources may include topic publications and subscriptions, memory that is held continuously, and initialising configuration parameters.
+
+Entering the `inactive` state from the `active` state requires running the node's `onDeactivated` method. See the `active` state description for more information.
+
+Entering the `inactive` state from the `fatal error` state requires running the node's `onReset` method. See the `fatal error` state description for more information.
+
+If an error occurs while in this state that cannot be handled by the node, then the transition to the `fatal error` state is taken.
+
+The node may also exit this state by being destroyed, in which case the `destroy` transition is taken.
+
+Implementation note: Causing the transition to the destroyed state can often be mapped to deletion of the node instance.
+
+### Active state
+
+This is the main state of the node's life cycle. While in this state, the node performs any processing, responds to service requests, reads and processes data, produces output, etc.
+
+Entering the `active` state requires running the node's `onActivated` method. This is used to prepare the node for execution. This may include acquiring resources that are only held while the node is actually active, such as access to hardware. Ideally, no preparation that requires significant time (such as lengthy hardware initialisation) should be performed in this callback.
+
+To shut down a node, it must first be deactivated. This will cause the `onDeactivate` method to be called, allowing the node to clean up any resources it only holds while active.
+
+If an error that cannot be handled by the node/system occurs in this state, the transition to the `fatal error` state is taken.
+
+### Fatal Error state
+
+Known errors that can occur in a node should be considered designed behaviour of the system and so should be handled by the node where the error occurred and, if necessary, the surrounding nodes. The reason for this is that the system knows best how to handle errors that occur in its behaviour.
+
+Any fatal errors (those which the node/system was not expecting and does not know how to handle) will cause a transition to the `fatal error` state. In this state, because the node does not know what to do in the presence of the error, it is not provided any execution time. Arriving data is not processed, service requests are not responded to, and no output is produced.
+
+On entering the `fatal error` state, the node's `onError` method must be executed. This provides the node with a chance to clean up resources it was using.
```

Comment:
I think there can be one fail state with well defined error codes. The controller can then make intelligent decisions based on the value of the code, and provide a human readable string to help diagnose the problem.

FailParam. "Could not parse value of parameter X"
FailHW "motor did not respond to command Q"
FailException. "Received alloc error when creating resource D"


Sent from my iPhone

> On Sep 18, 2015, at 8:09 AM, Sylvain Joyeux <notifications@github.com> wrote:
> 
> In articles/node_lifecycle.md:
> 
> > +
> > +This is the main state of the node's life cycle. While in this state, the node performs any processing, responds to service requests, reads and processes data, produces output, etc.
> > +
> > +Entering the `active` state requires running the node's `onActivated` method. This is used to prepare the node for execution. This may include acquiring resources that are only held while the node is actually active, such as access to hardware. Ideally, no preparation that requires significant time (such as lengthy hardware initialisation) should be performed in this callback.
> > +
> > +To shut down a node, it must first be deactivated. This will cause the `onDeactivate` method to be called, allowing the node to clean up any resources it only holds while active.
> > +
> > +If an error that cannot be handled by the node/system occurs in this state, the transition to the `fatal error` state is taken.
> > +
> > +### Fatal Error state
> > +
> > +Known errors that can occur in a node should be considered designed behaviour of the system and so should be handled by the node where the error occurred and, if necessary, the surrounding nodes. The reason for this is that the system knows best how to handle errors that occur in its behaviour.
> > +
> > +Any fatal errors (those which the node/system was not expecting and does not know how to handle) will cause a transition to the `fatal error` state. In this state, because the node does not know what to do in the presence of the error, it is not provided any execution time. Arriving data is not processed, service requests are not responded to, and no output is produced.
> > +
> > +On entering the `fatal error` state, the node's `onError` method must be executed. This provides the node with a chance to clean up resources it was using.
> Is that accurate? I'm just trying to nail down the argument and get an actionable proposal out of it.
> 
> Thank for for trying to bring the discussion upwards ;-)
> 
> It is actually not accurate. The states are important because they are a form of reporting so I definitely do not advocate for removing them. The current proposal only has an error state for fatal errors. I would prefer to bring the state machine closer to RTT on these states by having a separate Exception and FatalError states (I use the RTT names for simplicity) where:
> 
> Exception is the state the component ends in after an error it did not handle itself. It can be recovered, i.e. it is possible to transition from this state back to the Initialize state. This transition is where the component would e.g. itself try to recover a lost device if it knows how to do it.
> FatalError is the state the component ends in after a transition to a "stopped" state raised an exception (i.e. transition Active->Inactive, Inactive->Created, Active->Exception, Exception->Created). This is non-recoverable, as it is assumed that the component's internal state is inconsistent and that therefore the component/process should be brought down.
> In addition RTT has a RuntimeError state that means "I am active but in an error state". This is the state that I never saw used.
> 
> About your anecdotes: the library behaviour/design is usually related to how you assume your component should work. For instance, you close/reopen between command and configuration modes, OK. I would do it differently (not close connection & always mandatorily connect in configuration mode to set a known-good configuration), but both ways have pros and cons.
> 
> The bottom line is: it's not an issue, you can still do that. However, you would do all of it in the Initialize transition, and properly time out (transition to Error) if you don't manage to get proper configuration. I always push a full configuration to the devices because I never know who tinkered with them (e.g. I want a setup that can be reproduced no matter what)
> 
> When you say
> 
> but those would all be responses to arbitrary constraints of the system, which I think is bad.
> 
> Why is that ? The point of having a node is to standardize the behaviour of a non-standard object (the device) into something that is standard and understood by all. Node design best practices are always driven by the desired system design.
> 
> do the delay in an error state before telling the supervisor to try the transition again.
> 
> Again, this is a matter of system design. I would do the delay during the Created->Initialized transition since the component does not know when was the last time one was connected to the device in the first place.
> 
> —
> Reply to this email directly or view it on GitHub.
>